### PR TITLE
Make Object::_document RAII-safe

### PR DESCRIPTION
**Pull request by adamkewley**
Fixes nit

### Brief summary of changes

- Changes `Object::_document` to a `shared_ptr` to clarify memory handling

### Testing I've completed

- None, I'm assuming that the original implementation doesn't rely on the raw pointer having an unusual lifetime

### CHANGELOG.md (choose one)

- updated.

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3670)

### CODE REVIEW

**Review by tkuchida — APPROVED** [review]
LGTM :+1:

**Comment by adamkewley** [comment]
> LGTM 👍

Thanks for reading into this @tkuchida - I'll merge it after I do one last check on `setDocument`: one thing I'm realizing is that `setDocument`'s old semantics are purely referential/leaky whereas the new one assumes control over whatever it receives, which might alter behavior

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,12 @@ request related to the change, then we may provide the commit.
 
 This is not a comprehensive list of changes but rather a hand-curated collection of the more notable ones. For a comprehensive history, see the [OpenSim Core GitHub repo](https://github.com/opensim-org/opensim-core).
 
+Unreleased
```
[the remark this answers is not in the document]
Comment (aymanhab) [reply CHANGELOG.md]: Upcoming is 4.6, we're already changing APIs elsewhere. Sorry for the late comment, otherwise LGTM, thanks @adamkewley and @tkuchida

**Comment by adamkewley** [comment]
All uses of `setDocument` in the `opensim-core`  tree are fine. E.g.:

- `OpenSim/Tools/InverseDynamicsTool.cpp:624: setDocument(new XMLDocument(newFileName));`
- `OpenSim/Tools/InverseKinematicsTool.cpp:361: setDocument(new XMLDocument(newFileName));`
- `OpenSim/Tools/InverseKinematicsTool.cpp:395: setDocument(new XMLDocument(newFileName));`

Quick search through a spattering of open-source OpenSim API projects:

- `opensim-org/opensim-gui` contains no references to it (it does contain references to a `setDocument` symbol, but it's for `InstructionsTopComponent.java`, no `OpenSim::Model`
- `ComputationalBiomechanicsLab/opensim-creator` contains no references to it
- `tgeijten/scone-core` contains no references to it
- `rcnl-org/nmsm-core` contains no references to it
- `perfanalytics/pose2sim`: no references to it

So I think it's safe to assume that adding unconditional ownership to the `setDocument` API (as in, the previous version would always leak any overwritten `XMLDocument`) is *probably* ok... *probably*.

**Pull request merged**